### ISSUE TITLE
Pallas Tutorial outputs RESOURCE_EXHAUSTED (Reopen)

### ISSUE THREAD

**Issue by ClashLuke** [google/jax#20568]
### Description

#20008 persists. Opening a new issue, as I can't re-open the old one.

The same goes for the matmul code in the [design doc](https://github.com/google/jax/blob/85cb16936c39e9b25326c21cfed19a28fc9ca643/docs/pallas/design.md), after fixing the  typos to be

```PYTHON
@partial(jax.jit, static_argnames=["block_shape", "activation"])
def matmul(x, y, *, block_shape, activation):
    block_m, block_n, block_k = block_shape
    fused_matmul = pl.pallas_call(partial(matmul_kernel, block_k=block_k, activation=activation), grid=(2, 2),
                                  out_shape=jax.ShapeDtypeStruct((x.shape[0], y.shape[1],), jnp.float32),
                                  in_specs=[pl.BlockSpec(lambda i, j: (i, 0), (block_m, x.shape[1])),
                                            pl.BlockSpec(lambda i, j: (0, j), (y.shape[0], block_n))],
                                  out_specs=pl.BlockSpec(lambda i, j: (i, j), (block_m, block_n)))
    return fused_matmul(x, y)
  ```
    
Adding `backend="gpu"` doesn't resolve the issue, although I would've expected it to force pallas to use the GPU and thefore not run into shared memory issues.
 
Similarly, reducing the matrix size (as mentioned in the original post) to 128x128 masks the issue, but shouldn't be necessary on a DGX-H100.

### System info (python version, jaxlib version, accelerator, etc.)

Reproduced on two systems (DGX-H100, local PC), with and without docker, with 0.4.25 and 0.4.23. Attaching one set of versions below:

```
jax                           0.4.23
jax-triton                    0.1.4
jaxlib                        0.4.23+cuda12.cudnn89
nvidia-cublas-cu12            12.3.4.1
nvidia-cuda-cupti-cu12        12.3.101
nvidia-cuda-nvcc-cu12         12.3.107
nvidia-cuda-nvrtc-cu12        12.3.107
nvidia-cuda-runtime-cu12      12.3.101
nvidia-cudnn-cu12             8.9.7.29
nvidia-cufft-cu12             11.0.12.1
nvidia-cusolver-cu12          11.5.4.101
nvidia-cusparse-cu12          12.2.0.103
nvidia-ml-py                  12.535.133
nvidia-nccl-cu12              2.19.3
nvidia-nvjitlink-cu12         12.3.101
```

**Comment by ClashLuke** [comment]
Unfortunately, even when reducing the matrix size to 128x128, a plain matmul kernel yields a maximal relative difference of 4.6x with the matmul example from the getting-started codebase.

```PYTHON
import jax
from functools import partial
from jax.experimental import pallas as pl
from jax import numpy as jnp
import numpy as np

# jax.config.update("jax_default_matmul_precision", jax.lax.Precision.HIGHEST)

def matmul_kernel(x_ref, y_ref, z_ref):
    z_ref[...] = x_ref[...] @ y_ref[...]


def matmul(x: jax.Array, y: jax.Array):
    return pl.pallas_call(matmul_kernel, out_shape=jax.ShapeDtypeStruct((x.shape[0], y.shape[1]), x.dtype), grid=(2, 2),
                          in_specs=[pl.BlockSpec(lambda i, j: (i, 0), (x.shape[0] // 2, x.shape[1])),
                                    pl.BlockSpec(lambda i, j: (0, j), (y.shape[0], y.shape[1] // 2))],
                          out_specs=pl.BlockSpec(lambda i, j: (i, j), (x.shape[0] // 2, y.shape[1] // 2)))(x, y)


k1, k2 = jax.random.split(jax.random.key(0))
x = jax.random.normal(k1, (128, 128))
y = jax.random.normal(k2, (128, 128))

z0 = matmul(x, y)
jax.numpy.set_printoptions(linewidth=200)

z1 = x @ y

np.testing.assert_allclose(z0, z1)

# Precision: DEFAULT
# Mismatched elements: 16383 / 16384 (100%)
# Max absolute difference: 0.03544998
# Max relative difference: 4.6466436
#  x: array([[  1.119106,   6.592681,  -6.125449, ...,  -6.331781, -12.025826,  15.435001],
#        [  0.513045,  -4.678412,   6.029325, ...,   0.53062 ,  18.566563, -13.24138 ],
#        [ -1.497669,  -4.136811,  11.874069, ...,   2.772527,   0.57508 , -15.837502],...
#  y: array([[  1.116904,   6.59951 ,  -6.13182 , ...,  -6.336188, -12.039021,  15.440063],
#        [  0.513664,  -4.678648,   6.041401, ...,   0.529472,  18.572517, -13.24641 ],
#        [ -1.499167,  -4.129358,  11.879996, ...,   2.776294,   0.580626, -15.845534],...


# Precision: HIGHEST
# Mismatched elements: 11529 / 16384 (70.4%)
# Max absolute difference: 1.9073486e-05
# Max relative difference: 0.00039604
#  x: array([[  1.116904,   6.59951 ,  -6.131817, ...,  -6.336188, -12.039024,  15.440062],
#        [  0.513661,  -4.678648,   6.041399, ...,   0.52947 ,  18.57252 , -13.246411],
#        [ -1.499167,  -4.129362,  11.880001, ...,   2.776296,   0.580623, -15.845534],...
#  y: array([[  1.116904,   6.59951 ,  -6.13182 , ...,  -6.336188, -12.039021,  15.440063],
#        [  0.513664,  -4.678648,   6.041401, ...,   0.529472,  18.572517, -13.24641 ],
#        [ -1.499167,  -4.129358,  11.879996, ...,   2.776294,   0.580626, -15.845534],...
```

**Comment by ClashLuke** [comment]
Is the tutorial meant for TPU?

It appears that "shared memory" references cuda's shared memory, not CPU shared memory as I thought initially. Reducing the tile size to 16x16 (with K=full_dim) allows running the code normally.

```PYTHON
def matmul(x: jax.Array, y: jax.Array, tile_n: int = 16, tile_m: int = 16):
    return pl.pallas_call(matmul_kernel, out_shape=jax.ShapeDtypeStruct((x.shape[0], y.shape[1]), x.dtype),
                          grid=(x.shape[0] // tile_n, y.shape[1] // tile_m),
                          in_specs=[pl.BlockSpec(lambda i, j: (i, 0), (tile_n, x.shape[1])),
                                    pl.BlockSpec(lambda i, j: (0, j), (y.shape[0], tile_m))],
                          out_specs=pl.BlockSpec(lambda i, j: (i, j), (tile_n, tile_m)))(x, y)


k1, k2 = jax.random.split(jax.random.key(0))
x = jax.random.normal(k1, (512, 512), dtype=jnp.float32)
y = jax.random.normal(k2, (512, 512), dtype=jnp.float32)
```

So, it seems everything works as intended, and I'm excited to get building with pallas. A tutorial without errors would still be nice to have.

**Comment by domluna** [comment]
@ClashLuke this might be helpful to you

https://github.com/google/jax/discussions/20509 I got the matmul working on my GPU (GTX 3090) with the linked code. The out-of-memory error seems to be related to going over the static shared memory limit for the GPU, which is around 100KB depending on your GPU. So if I set the block size to 128x128 and the type to bfloat16, it now works fine.

In [30]: 128 * 128 * 2 * 2
Out[30]: 65536

In [31]: 128 * 128 * 2 * 3
Out[31]: 98304

Depending on how many matrices you need, you should do some quick math to make sure it's within the memory limit. Pallas might be doing optimizations under the hood as well. Note that this seems to be only relevant for tensor core kernels, so if you have a kernel that does some broadcast operation like x + 1.3, the block size doesn't matter, apart from performance implications because this uses dynamic shared memory by default.
